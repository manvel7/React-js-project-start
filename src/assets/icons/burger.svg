<svg width="22" height="16" viewBox="0 0 22 16" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M0 0.5C0 0.223858 0.223858 0 0.5 0H21.5C21.7761 0 22 0.223858 22 0.5V1.5C22 1.77614 21.7761 2 21.5 2H0.5C0.223857 2 0 1.77614 0 1.5V0.5Z" fill="#001254"/>
<path d="M0 7.5C0 7.22386 0.223858 7 0.5 7H21.5C21.7761 7 22 7.22386 22 7.5V8.5C22 8.77614 21.7761 9 21.5 9H0.5C0.223857 9 0 8.77614 0 8.5V7.5Z" fill="#001254"/>
<path d="M0.5 14C0.223858 14 0 14.2239 0 14.5V15.5C0 15.7761 0.223857 16 0.5 16H21.5C21.7761 16 22 15.7761 22 15.5V14.5C22 14.2239 21.7761 14 21.5 14H0.5Z" fill="#001254"/>
</svg>
