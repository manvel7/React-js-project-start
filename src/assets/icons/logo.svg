<svg width="20" height="25" viewBox="0 0 20 25" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M11.0379 25V22.2294C11.7204 22.4114 12.5807 22.5718 13.4953 22.5718C14.3653 22.5718 15.2842 22.4267 16.1442 22.017C19.0324 20.6415 19.9365 16.9139 19.9735 16.7559C20.0829 16.2895 19.8459 15.8114 19.4042 15.6075C19.2546 15.5385 15.7061 13.9313 12.8177 15.307C12.111 15.6436 11.5233 16.1211 11.0379 16.6526V12.5993C11.5588 12.7135 12.181 12.8049 12.8409 12.8049C13.6627 12.8049 14.542 12.6632 15.3535 12.2477C17.851 10.9693 18.5437 7.69365 18.572 7.55482C18.6676 7.08557 18.4165 6.61442 17.969 6.42316C17.8366 6.3666 14.698 5.0523 12.2006 6.33079C11.7585 6.55711 11.3732 6.84616 11.0379 7.16919V4.79567C11.0716 3.74499 10.9074 2.20411 9.85043 1.16808C8.20949 -0.44029 5.32562 0.0596159 5.00228 0.121875C4.58857 0.201565 4.26522 0.518491 4.18392 0.923991C4.1204 1.24098 3.61044 4.06768 5.25131 5.67598C6.22895 6.63415 7.6475 6.84399 8.7032 6.84399C8.79225 6.84399 8.87874 6.8425 8.96212 6.83979V12.676C8.47665 12.1446 7.88906 11.667 7.18231 11.3304C4.29428 9.95493 0.745455 11.5619 0.595788 11.6309C0.154121 11.8348 -0.0829427 12.3128 0.0265225 12.7793C0.0635414 12.9373 0.967634 16.6649 3.8558 18.0404C4.71609 18.4501 5.6345 18.5952 6.50469 18.5952C7.41923 18.5952 8.27952 18.4348 8.96212 18.2528V24.9999L11.0379 25ZM13.7249 17.1371C15.0357 16.5128 16.6901 16.8415 17.6734 17.1415C17.3062 18.0853 16.5474 19.563 15.2369 20.1871C13.9274 20.8109 12.2743 20.4832 11.2909 20.1833C11.6607 19.2388 12.4229 17.7572 13.7249 17.1371ZM13.1617 8.13435C14.159 7.62379 15.4376 7.80087 16.2721 8.012C15.9719 8.80462 15.3897 9.93383 14.3926 10.4443C13.3953 10.9549 12.1166 10.7778 11.2821 10.5666C11.5822 9.77405 12.1644 8.64484 13.1617 8.13435ZM4.76301 16.2106C3.45233 15.5864 2.69347 14.1082 2.32647 13.1651C3.31034 12.8648 4.96464 12.5364 6.27504 13.1605C7.58571 13.7847 8.34457 15.2628 8.7115 16.2059C7.72763 16.5063 6.07355 16.8348 4.76301 16.2106ZM8.95666 4.80293C8.22555 4.83168 7.23642 4.74426 6.71906 4.23724C6.20723 3.7355 6.11748 2.76329 6.14495 2.04412C6.87979 2.01719 7.87093 2.10516 8.38262 2.60669C8.89521 3.10917 8.98447 4.08349 8.95666 4.80293Z" fill="#FEFEFE"/>
</svg>
