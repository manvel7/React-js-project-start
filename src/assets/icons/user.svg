<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<g opacity="0.5">
<path d="M20.3973 17.933C18.5357 20.558 15.4687 22.2857 12 22.2857C8.53125 22.2857 5.46429 20.558 3.60268 17.933C3.95089 15.4286 4.96875 13.2054 7.24554 12.8973C8.42411 14.183 10.125 15 12 15C13.875 15 15.5759 14.183 16.7545 12.8973C19.0312 13.2054 20.0491 15.4286 20.3973 17.933ZM17.1429 8.57143C17.1429 11.4107 14.8393 13.7143 12 13.7143C9.16071 13.7143 6.85714 11.4107 6.85714 8.57143C6.85714 5.73214 9.16071 3.42857 12 3.42857C14.8393 3.42857 17.1429 5.73214 17.1429 8.57143ZM24 12C24 5.37054 18.6295 0 12 0C5.37054 0 0 5.37054 0 12C0 18.6161 5.37054 24 12 24C18.6429 24 24 18.6027 24 12Z" fill="#FEFEFE"/>
</g>
</svg>
