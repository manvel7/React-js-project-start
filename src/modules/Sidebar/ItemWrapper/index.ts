import SidebarItemWrapper from "./SidebarItemWrapper"

export default SidebarItemWrapper
