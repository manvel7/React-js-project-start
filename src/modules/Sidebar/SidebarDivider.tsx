import {HTMLAttributes} from "react"
import styled from "styled-components"

interface IProps extends HTMLAttributes<HTMLHRElement> {}

const SidebarDivider = styled.hr<IProps>`
  border-color: ${({theme}) => theme.palette.background.white};
  margin: 1.6rem auto;
  opacity: 0.3;
  width: calc(100% - 4rem);
`

export default SidebarDivider
