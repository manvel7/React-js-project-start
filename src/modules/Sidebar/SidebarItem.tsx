import React, {FunctionComponent, ReactNode} from "react"
import SidebarItemWrapper, {
  IProps as IWrapperProps
} from "./ItemWrapper/SidebarItemWrapper"

interface IProps extends IWrapperProps {
  icon: ReactNode
  title?: string
}

const SidebarItem: FunctionComponent<IProps> = (props) => {
  const {icon, title, ...rest} = props

  return (
    <SidebarItemWrapper {...rest}>
      <div className="icon">{icon}</div>
      {title && (
        <div className="text">
          <h6>{title}</h6>
        </div>
      )}
    </SidebarItemWrapper>
  )
}

export default SidebarItem
