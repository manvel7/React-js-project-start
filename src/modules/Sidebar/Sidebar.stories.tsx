import React from "react"
import {Meta, Story} from "@storybook/react/types-6-0"
import {ReactComponent as Burger} from "../../assets/icons/burger.svg"
import {ReactComponent as Logo} from "../../assets/icons/logo.svg"
import {ReactComponent as Scout} from "../../assets/icons/scout.svg"
import {ReactComponent as User} from "../../assets/icons/user.svg"
import Sidebar, {IProps} from "./Sidebar"
import {SidebarDivider, SidebarItem} from "./index"

const meta: Meta<IProps> = {
  title: "Modules/Sidebar",
  component: Sidebar
}

const SidebarStory: Story<IProps> = (args) => (
  <Sidebar {...args} ToggleIcon={Burger}>
    <SidebarItem icon={<Logo />} title="Logo" displayPosition="top" />
    <SidebarItem icon={<Scout />} title="Scout" />
    <SidebarDivider />
    <SidebarItem icon={<Scout />} title="Scout" />
    <SidebarItem icon={<Scout />} title="Scout (auto)" displayPosition="auto" />
    <SidebarItem icon={<User />} title="Profile" displayPosition="bottom" />
  </Sidebar>
)

const Default = SidebarStory.bind({})

export {Default}

export default meta
