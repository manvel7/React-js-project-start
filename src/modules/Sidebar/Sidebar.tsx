import React, {
  FunctionComponent,
  ComponentType,
  HTMLAttributes,
  useState,
  MouseEvent
} from "react"
import clsx from "clsx"
import SidebarDivider from "./SidebarDivider"
import SidebarItem from "./SidebarItem"
import SidebarWrapper from "./SidebarWrapper"

export interface IProps extends HTMLAttributes<HTMLDivElement> {
  ToggleIcon: ComponentType
  activeAppName?: string
}

const Sidebar: FunctionComponent<IProps> = (props) => {
  const {children, activeAppName, ToggleIcon} = props
  const [isExpanded, setIsExpanded] = useState<boolean>(false)

  const onMenuToggle = () => {
    setIsExpanded((prev) => !prev)
  }

  const sidebarItems = React.Children.map(children, (item: any) => {
    if (
      typeof item !== "object" ||
      (item.type !== SidebarItem && item.type !== SidebarDivider)
    ) {
      throw new TypeError(
        "Sidebar children should only contain SidebarItem and SidebarDivider elements"
      )
    }

    if (item.type === SidebarDivider) {
      return item
    }

    return React.cloneElement(item, {
      className: clsx(item.props.className),
      onClick(event: MouseEvent<HTMLDivElement>) {
        setIsExpanded(false)

        if (item.props.onClick) {
          item.props.onClick(event)
        }
      }
    })
  })

  return (
    <SidebarWrapper
      className={clsx({
        expanded: isExpanded
      })}>
      <div className="mobile-heading">
        <div className="burger-button" onClick={onMenuToggle}>
          <ToggleIcon />
        </div>
        <h4>{activeAppName}</h4>
      </div>
      {sidebarItems}
    </SidebarWrapper>
  )
}

Sidebar.defaultProps = {
  activeAppName: "Linda Forest"
}

export default Sidebar
