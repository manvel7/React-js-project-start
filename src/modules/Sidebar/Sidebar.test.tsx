import React from "react"
import {render} from "@testing-library/react"
import {SvgMock as Burger} from "../../assets/icons/burger.svg"
import {SvgMock as Logo} from "../../assets/icons/logo.svg"
import {SvgMock as Scout} from "../../assets/icons/scout.svg"
import {SvgMock as User} from "../../assets/icons/user.svg"
import {UiProvider} from "../../styles"
import Sidebar from "./Sidebar"
import SidebarDivider from "./SidebarDivider"
import SidebarItem from "./SidebarItem"

describe("Sidebar", () => {
  it("should match snapshot", () => {
    const result = render(
      <UiProvider>
        <Sidebar ToggleIcon={Burger}>
          <SidebarItem icon={<Logo />} title="Logo" displayPosition="top" />
          <SidebarItem icon={<Scout />} title="Scout" />
          <SidebarDivider />
          <SidebarItem
            icon={<Scout />}
            title="Scout (auto)"
            displayPosition="auto"
          />
          <SidebarItem
            icon={<User />}
            title="Profile"
            displayPosition="bottom"
          />
        </Sidebar>
      </UiProvider>
    )
    expect(result.container.firstChild).toMatchSnapshot()
  })
})
