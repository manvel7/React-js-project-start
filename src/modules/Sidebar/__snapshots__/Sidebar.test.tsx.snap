// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should match snapshot 1`] = `
.c3 {
  border-color: #FEFEFE;
  margin: 1.6rem auto;
  opacity: 0.3;
  width: calc(100% - 4rem);
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 2.4rem;
}

.c1:hover {
  background-color: #000A62;
  cursor: pointer;
}

.c1 .icon,
.c1 .text {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 .icon {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 6.4rem;
  width: 6.4rem;
}

.c1 .text {
  padding-left: 1.6rem;
  white-space: nowrap;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2:hover {
  background-color: #000A62;
  cursor: pointer;
}

.c2 .icon,
.c2 .text {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 .icon {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 6.4rem;
  width: 6.4rem;
}

.c2 .text {
  padding-left: 1.6rem;
  white-space: nowrap;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: auto;
  margin-top: auto;
}

.c4:hover {
  background-color: #000A62;
  cursor: pointer;
}

.c4 .icon,
.c4 .text {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 .icon {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 6.4rem;
  width: 6.4rem;
}

.c4 .text {
  padding-left: 1.6rem;
  white-space: nowrap;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: auto;
}

.c5:hover {
  background-color: #000A62;
  cursor: pointer;
}

.c5 .icon,
.c5 .text {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c5 .icon {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  height: 6.4rem;
  width: 6.4rem;
}

.c5 .text {
  padding-left: 1.6rem;
  white-space: nowrap;
}

.c0 {
  background-color: #001254;
  bottom: 0;
  color: #FEFEFE;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  left: 0;
  overflow-x: hidden;
  overflow-y: auto;
  position: fixed;
  top: 0;
  -webkit-transition: width 0.4s ease-in-out;
  transition: width 0.4s ease-in-out;
  width: 6.4rem;
  z-index: 10;
}

.c0:hover {
  -webkit-transition-duration: 0.25s;
  transition-duration: 0.25s;
  width: 25.6rem;
}

.c0 .mobile-heading {
  display: none;
  -webkit-flex: 0 0 5.6rem;
  -ms-flex: 0 0 5.6rem;
  flex: 0 0 5.6rem;
  line-height: 5.6rem;
  position: relative;
  text-align: center;
}

.c0 .mobile-heading h4 {
  line-height: 5.6rem;
}

.c0 .mobile-heading .burger-button {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  bottom: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0 2.4rem;
  position: absolute;
  top: 0;
}

.c0 .mobile-heading .burger-button svg path {
  fill: #FEFEFE;
}

@media (max-width:575px) {
  .c0 {
    bottom: initial;
    right: 0;
    width: initial;
    height: 5.6rem;
    -webkit-transition: height 0.4s ease-in-out;
    transition: height 0.4s ease-in-out;
    overflow-y: hidden;
  }

  .c0:hover {
    width: initial;
  }

  .c0.expanded {
    height: 100vh;
    overflow-y: auto;
  }

  .c0 .mobile-heading {
    display: block;
  }
}

<div
  class="c0"
>
  <div
    class="mobile-heading"
  >
    <div
      class="burger-button"
    />
    <h4>
      Linda Forest
    </h4>
  </div>
  <div
    class="c1"
  >
    <div
      class="icon"
    />
    <div
      class="text"
    >
      <h6>
        Logo
      </h6>
    </div>
  </div>
  <div
    class="c2"
  >
    <div
      class="icon"
    />
    <div
      class="text"
    >
      <h6>
        Scout
      </h6>
    </div>
  </div>
  <hr
    class="c3"
  />
  <div
    class="c4"
  >
    <div
      class="icon"
    />
    <div
      class="text"
    >
      <h6>
        Scout (auto)
      </h6>
    </div>
  </div>
  <div
    class="c5"
  >
    <div
      class="icon"
    />
    <div
      class="text"
    >
      <h6>
        Profile
      </h6>
    </div>
  </div>
</div>
`;
